src/reader/warming.rs

```diff
@@ -0,0 +1,353 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex, Weak},
+    thread::JoinHandle,
+    time::Duration,
+};
+
+use crate::{Executor, Searcher, SearcherGenerationToken, SearcherIndexGeneration, TantivyError};
+
+pub const GC_INTERVAL: Duration = Duration::from_secs(1);
+
+/// `Warmer` can be used to maintain segment-level state e.g. caches.
+///
+/// They must be registered with the [IndexReaderBuilder].
+pub trait Warmer: Sync + Send {
+    /// Perform any warming work using the provided [Searcher].
+    fn warm(&self, searcher: &Searcher) -> crate::Result<()>;
+
+    /// Discard internal state for any [SearcherIndexGeneration] not provided.
+    fn garbage_collect(&self, live_generations: &HashSet<SearcherIndexGeneration>);
+}
+
+/// Warming-related state with interior mutability.
+#[derive(Clone)]
+pub(crate) struct WarmingState(Arc<Mutex<WarmingStateInner>>);
+
+impl WarmingState {
+    pub fn new(num_warming_threads: usize, warmers: Vec<Weak<dyn Warmer>>) -> crate::Result<Self> {
+        Ok(Self(Arc::new(Mutex::new(WarmingStateInner {
+            num_warming_threads,
+            warmers,
+            searcher_generations: HashMap::new(),
+            gc_thread: None,
+        }))))
+    }
+
+    /// Start tracking a new generation of [Searcher], and [Warmer::warm] it if there are active warmers.
+    ///
+    /// A background GC thread for [Warmer::garbage_collect] calls is uniquely created if there are active warmers.
+    pub fn new_searcher_generation(&self, searcher: &Searcher) -> crate::Result<()> {
+        self.0
+            .lock()
+            .unwrap()
+            .new_searcher_generation(searcher, &self.0)
+    }
+
+    #[cfg(test)]
+    fn gc_maybe(&self) -> bool {
+        self.0.lock().unwrap().gc_maybe()
+    }
+}
+
+struct WarmingStateInner {
+    num_warming_threads: usize,
+    warmers: Vec<Weak<dyn Warmer>>,
+    searcher_generations: HashMap<SearcherIndexGeneration, Vec<SearcherGenerationToken>>,
```

Comment:
I am confused. Why do we need `Vec<SearcherGenerationToken>` here?

Isn't `HashMap<SearcherIndexGeneration, SearcherGenerationToken>` suficient ?
I think all `SearcherGenerationToken` for a given SearcherIndexGeneration are identical `Weak` reference coupled to the same set of `Arc` clones.

Reply:
If the above is correct, the HashMap is useless too...
It would become 

`Vec<SearcherGenerationToken>`

Reply:
That's true. They will be identical though...

In your use case, your warmed data is keyed by `SegmentId` only and it might not really matter much... But let's dig a little bit.

One could imagine a Warmer that grabs updated values from an external source (say price from a SQL db), grabs the docid->someprimaryid from a fast field and creates a column `docid -> newprice using the `Warmer`. 
The Searcher is supposed to a consistent view of the index throughout its lifetime.

Calling reload manually, even without any change in the index is meaningful in that world: it updates the price.

So we want the API to allow:
- Warming stuff, and keeping things per `(Segment, DeleteOpstamp)`
- Warming stuff, and keeping things per `Searcher generation`.

Reply:
> Calling reload manually, even without any change in the index is meaningful in that world

`warm()` will indeed get called -- it's a redundant `garbage_collect()` call that can be avoided when one of those `SearcherGenerationToken`s that point to an == `SearcherIndexGeneration` are dropped. But I think not that big a deal to have a redundant call, as we always provide the set of live `SearcherIndexGeneration` to test state against.